src/provider/index.ts

```diff
@@ -54,37 +64,70 @@ export class Main {
-            [APITopic.TOGGLE_NOTIFICATION_CENTER]: this.toggleNotificationCenter.bind(this)
+            [APITopic.TOGGLE_NOTIFICATION_CENTER]: this.toggleNotificationCenter.bind(this),
+            [APITopic.REGISTER_CLIENT]: this.registerClient.bind(this),
+            [APITopic.UNREGISTER_CLIENT]: this.unregisterClient.bind(this)
         });
 
         this._store.onAction.add((action: RootAction) => {
-            if (action.type === Action.REMOVE) {
-                // Send notification closed event to uuid with the context.
-                action.notifications.forEach((notification: StoredNotification) => {
-                    const target: Identity = notification.source;
-                    const event: NotificationClosedEvent = {type: 'notification-closed', notification: notification.notification};
-
-                    this._apiHandler.dispatchClientEvent(target, event);
-                });
-            } else if (action.type === Action.CLICK_BUTTON) {
-                const {notification, buttonIndex} = action;
-                const target: Identity = notification.source;
-                const event: NotificationButtonClickedEvent = {
-                    type: 'notification-button-clicked',
-                    notification: notification.notification,
-                    buttonIndex
-                };
-                this._apiHandler.dispatchClientEvent(target, event);
-            } else if (action.type === Action.CLICK_NOTIFICATION) {
-                const {notification, source} = action.notification;
-                const event: NotificationClickedEvent = {type: 'notification-clicked', notification};
+            let client: Identity | undefined;
 
-                // Send notification clicked event to uuid with the context.
-                this._apiHandler.dispatchClientEvent(source, event);
+            if (action.type === Action.CLICK_BUTTON || action.type === Action.CLICK_NOTIFICATION) {
+                client = action.notification.source;
+            } else if (action.type === Action.REMOVE && action.notifications.length > 0) {
+                client = action.notifications[0].source;
+            } else if (action.type === Action.DISPATCH_DEFERRED_EVENTS) {
+                client = action.target;
             }
-        });
 
-        console.log('Service Initialised');
+            if (client !== undefined) {
+                const connected: boolean = this._apiHandler.isAppConnected(client!.uuid);
+                const deferDispatchOrDiscard = (target: Identity, event: NotificationEvent): void => {
+                    if (this.interestMap.has(event.type, target)) {
+                        if (connected) {
+                            this._apiHandler.dispatchAppEvent(target.uuid, event);
+                        } else {
+                            this._store.dispatch({type: Action.DEFER_EVENT_DISPATCH, target, event});
+                            clientInfoStorage.getItem(target.uuid).then(item => {
```

Comment:
An app could theoretically be running already in this case, right? App starts, registers event listeners, app navigates to a new page that doesn't register event listeners...

Reply:
Yes, but if the current page doesn't register any listeners and the previous page was the only page that has the listeners for this event type, either;
1) defer them (and dispatch once the previous page is back up and performs handshake)
2) ignore them (if the previous page unregistered before the transition)